src/main/kotlin/com/demonwav/mcdev/facet/MinecraftFacet.kt

```diff
@@ -0,0 +1,247 @@
+
+            if (good) {
+                return action(abstractModule)
+            }
+        }
+        return null
+    }
+
+    @Contract(pure = true)
+    fun isEventGenAvailable() = modules.keys.any { it.isEventGenAvailable }
+
+    @Contract(pure = true)
+    fun shouldShowPluginIcon(element: PsiElement?) = modules.values.any { it.shouldShowPluginIcon(element) }
+
+    @Contract(pure = true)
+    fun getIcon(): Icon? {
+        val iconCount = modules.keys.count { it.hasIcon() }
+        if (iconCount == 0) {
+            return null
+        } else if (iconCount == 1) {
+            return modules.keys.firstOrNull { it.hasIcon() }?.icon
+        } else if (
+            iconCount == 2 &&
+            modules.containsKey(SpongeModuleType) &&
+            modules.containsKey(ForgeModuleType)
+        ) {
+            return PlatformAssets.SPONGE_FORGE_ICON
+        } else if (modules.size > 0) {
+            return PlatformAssets.MINECRAFT_ICON
+        } else {
+            return null
+        }
+    }
+
+    fun findFile(path: String, type: SourceType): VirtualFile? {
+        val roots = roots[type]
+        for (root in roots) {
+            return root.findFileByRelativePath(path) ?: continue
+        }
+        return null
+    }
+
+    companion object {
+        @JvmField
+        val ID = FacetTypeId<MinecraftFacet>(TYPE_ID)
+
+        @JvmStatic
+        val facetType
+            get() = FacetTypeRegistry.getInstance().findFacetType(ID) as MinecraftFacetType
+
+        @JvmStatic
+        fun getInstance(module: Module) = FacetManager.getInstance(module).getFacetByType(ID)
+
+        @JvmStatic
+        fun getChildInstances(module: Module): Set<MinecraftFacet> {
+            val instance = getInstance(module)
+            if (instance != null) {
+                return setOf(instance)
+            }
+
+            val manager = ModuleManager.getInstance(module.project)
+            val result = mutableSetOf<MinecraftFacet>()
+
+            for (m in manager.modules) {
+                val path = manager.getModuleGroupPath(m) ?: continue
+                val namedModule = manager.findModuleByName(path.last()) ?: continue
+
+                if (namedModule != module) {
+                    continue
+                }
+
+                val facet = getInstance(m) ?: continue
+                result.add(facet)
+            }
+            return result
+        }
+
+        @JvmStatic
+        fun <T : AbstractModule> getInstance(module: Module, type: AbstractModuleType<T>): T? {
+            val instance = getInstance(module) ?: return null
+            return instance.getModuleOfType(type)
+        }
+
+        @JvmStatic
+        fun <T : AbstractModule> getInstance(module: Module, vararg types: AbstractModuleType<*>): T? {
+            val instance = getInstance(module) ?: return null
+            @Suppress("UNCHECKED_CAST")
+            return types.asSequence().mapNotNull { instance.getModuleOfType(it) }.firstOrNull() as? T
```

Comment:
[`mapFirstNotNull`](https://github.com/minecraft-dev/MinecraftDev/blob/2016.3/src/main/kotlin/com/demonwav/mcdev/util/Util.kt#L50) would be more efficient here, it will immediately cancel once the first has been found.